### PR TITLE
IRGen: emit type metadata and (value) witness tables lazily.

### PR DESCRIPTION
**Pull request by eeckstein**
This gives big code size wins for unused types and also for types, which are never used in a generic context.
Also it reduces the amount of symbols in the symbol table.
The size wins heavily depend on the project. I have seen binary size reductions from 0 to 20% on real world projects.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci Please test

File: lib/IRGen/GenDecl.cpp [hunk 1]
```diff
@@ -59,27 +59,39 @@
...
-    return false;
-  case FormalLinkage::PublicNonUnique:
-  case FormalLinkage::HiddenNonUnique:
-    return true;
+  switch (Nominal->getKind()) {
+    case DeclKind::Enum:
+    case DeclKind::Struct:
+      break;
+    default:
+      // Keep all metadata for classes, because a class can be instanciated by
```
Comment (swiftix) [inline lib/IRGen/GenDecl.cpp]: spelling: it should be `instantiated`

**Comment by eeckstein** [comment]
@swift-ci Please benchmark

**Comment by eeckstein** [comment]
@swift-ci Please benchmark

File: lib/IRGen/GenDecl.cpp (hunk 1 above)
Comment (jrose-apple) [inline lib/IRGen/GenDecl.cpp]: NSClassFromString can also find classes on Apple platforms.

**Comment by jrose-apple** [comment]
Can't we end up using an unreferenced witness table via `as?`? Or will that count as a reference?

**Comment by dabrahams** [comment]
on Fri Mar 10 2017, eeckstein <notifications-AT-github.com> wrote:

> I have seen binary size reductions from 0 to 20% on real world
> projects.

That's YUGE!

**Comment by eeckstein** [comment]
> Can't we end up using an unreferenced witness table via as?? Or will that count as a reference?

@jrose-apple Whenever a metatype of a type is referenced (which means: may escape) all it's conformances are considered as referenced. This should cover the dynamic conformance check with as?

**Comment by eeckstein** [comment]
@slavapestov Here it is: rdar://problem/31045008

**Comment by jrose-apple** [comment]
Oh, right, I keep forgetting that putting something into a box references its metatype.

…does this mean that classes are always considered to have escaping metatypes, since the metatype is referenced during allocation?

**Comment by jrose-apple** [comment]
Wait, hang on, what if *another* compilation unit puts the thing into the box, but the conformance is declared in this compilation unit? (And does not escape.)

```swift
// A.swift
public struct Foo {}
public func boxedFoo() -> Any { return Foo() }
```

```swift
// B.swift
private protocol Bar {}
extension Foo: Bar {}
func test() {
  if let bar = boxedFoo() as? Bar {
    print("good")
  } else {
    print("bad")
  }
}
```

(these are separate but non-resilient modules)

**Comment by eeckstein** [comment]
> Wait, hang on, what if another compilation unit puts the thing into the box, but the conformance is declared in this compilation unit? (And does not escape.)

We only emit metatypes/conformances lazily for types which are local to the compilation unit (internal for wmo and private for non-wmo).

> …does this mean that classes are always considered to have escaping metatypes, since the metatype is referenced during allocation?

Yes. And also we even don't try to lazily emit class metatypes

**Comment by jrose-apple** [comment]
If I'm reading the check correctly, it's seeing whether the *witness table* is local to the compilation unit, not the base type. That's not the same thing.

I was more worried about not lazily emitting *conformances* on classes, but I guess we can do that later.

**Comment by eeckstein** [comment]
As I said, class conformances will always be emitted.

About witness table linkage: the SILLinkage of a witness table is derived from the minimum visibility of the conforming type and the protocol. So either the metatype of a compilation-unit local type does not escape or the protocol itself is not visible outside the compilation unit.

**Comment by jrose-apple** [comment]
> So either the metatype of a compilation-unit local type does not escape or the protocol itself is not visible outside the compilation unit.

Right, but that doesn't remove the need to emit the conformance; see my example above.

**Comment by eeckstein** [comment]
Oops, you are right. I should check the type's visibility and not the witness table linkage.
I'll fix that

**Pull request merged**